### ISSUE TITLE
Inefficient recompilation

### ISSUE THREAD

**Issue by felixfbecker** [sourcegraph/javascript-typescript-langserver#113]
Right now, `ProjectConfiguration.reset()` unsets everything including the `Program` and `TypeScriptService`. Then the next call to `init()` will initialize everything from scratch. This is called by `refreshFileTree()`, which is called for example after dependency installation.

Afaik, this is not the intended usage. When creating a `Program`, you are supposed to pass in an old `Program`, and SourceFiles etc that can be reused and didn't change can be reused. I don't think the LanguageService should be destroyed.

**Comment by alexsaveliev** [comment]
>  When creating a Program, you are supposed to pass in an old Program, and SourceFiles etc that can be reused and didn't change can be reused.

Yeah, but the devil is in the detail. One cannot simply pass `ts.createCompilerHost(...)` to `ts.createProgram(..)`, you should construct proper instance of `CompilerHost` object which is a large part of `LanguageService::getProgram()`

Anyway, I agree that we need to act at some point
